integration_tests/src/main/python/parquet_pyarrow_test.py

```diff
@@ -0,0 +1,255 @@
+    py_key_value_pair_list = pyarrow_map_scalar.as_py()
+    if py_key_value_pair_list is None:
+        return None
+    else:
+        return dict(py_key_value_pair_list)
+
+
+def _convert_pyarrow_column_to_cpu_object(pyarrow_column, pa_type):
+    """ convert pyarrow column to a column contains CPU type objects"""
+    if isinstance(pa_type, pa.StructType):
+        # struct column
+        return [_convert_pyarrow_struct_object_to_cpu_object(v) for v in pyarrow_column]
+    elif isinstance(pa_type, pa.ListType):
+        # list column
+        return [_convert_pyarrow_list_object_to_cpu_object(v) for v in pyarrow_column]
+    elif isinstance(pa_type, pa.MapType):
+        # map column
+        return [_convert_pyarrow_map_object_to_cpu_object(v) for v in pyarrow_column]
+    else:
+        # other type column
+        return [v.as_py() for v in pyarrow_column]
+
+
+def _gen_row_from_columnar_cpu_data(column_names, columnar_data, row_idx):
+    """Gen a Pyspark Row from columnar data for a row index"""
+    key_value_pair_list = [(column_names[col], columnar_data[col][row_idx]) for col in range(len(column_names))]
+    return Row(**dict(key_value_pair_list))
+
+
+def _convert_pyarrow_table_to_cpu_object(pa_table):
+    """
+    Convert pyarrow table to cpu type data to reuse the `_assert_equal(cpu, gpu)` in asserts.py
+     e.g.: convert pyarrow map_scalar to Pyspark Row
+    """
+    columnar_data = []
+    for pa_column, pa_type in zip(pa_table.columns, pa_table.schema.types):
+        converted_column = _convert_pyarrow_column_to_cpu_object(pa_column, pa_type)
+        columnar_data.append(converted_column)
+    cpu_data = [
+        _gen_row_from_columnar_cpu_data(pa_table.schema.names, columnar_data, row_idx)
+        for row_idx in range(pa_table.num_rows)
+    ]
+    return cpu_data
+
+
+def assert_equal(cpu_adapted_from_pyarrow, gpu):
+    """Verify that the result from the pyarrow(already adapted to CPU data format) and the GPU are equal"""
+    try:
+        _assert_equal(cpu_adapted_from_pyarrow, gpu, float_check=get_float_check(), path=[],
+                      cpu_data_adapted_from="pyarrow")
+    except:
+        sys.stdout.writelines(difflib.unified_diff(
+            a=[f"{x}\n" for x in cpu_adapted_from_pyarrow],
+            b=[f"{x}\n" for x in gpu],
+            fromfile='pyarrow OUTPUT',
+            tofile='GPU OUTPUT'))
+        raise
+
+
+def assert_gpu_and_pyarrow_are_compatible(base_write_path, gen_list, conf={}):
+    """
+    Test scenarios:
+      Write by pyarrow, test readings on pyarrow and GPU
+      Write by GPU, test readings on pyarrow and GPU
+    The data is collected back to the driver and compared here, so be
+    careful about the amount of data returned.
+    """
+    conf = _prep_incompat_conf(conf)
+    pyarrow_parquet_path = base_write_path + "/pyarrow"
+    gpu_parquet_path = base_write_path + "/gpu"
+
+    def write_on_pyarrow():
+        pa_table = _gen_pyarrow_table(gen_list)
```

Comment:
This is more of a question than a suggestion. Spark has a toPandas API that will convert a dataframe to pandas. Even though it is really convoluted to generate the data, send it to the spark cluster just to pull it back here. It might make some of the code simpler, but the functions it would replace are not that big and complicated anyways.

Reply:
I tried the following writing method:
``` python
    def write_on_pyarrow2(spark, data_gen):
        df = gen_df(spark, data_gen, 2048)
        pdf = df.toPandas()
        table = pa.Table.from_pandas(pdf)
        pa_pq.write_table(table, pyarrow_parquet_path)
```

Some cases failed but previously they are OK.

Reply:
As Mithun mentioned, Pandas has limitations:
- date range limitation
```py
import pandas as pd
pd.Timestamp('8543-12-31')
```
```
pandas._libs.tslibs.np_datetime.OutOfBoundsDatetime: Out of bounds nanosecond timestamp: 8543-12-31 00:00:00
```
- Conversion from Pandas dataframe
refer to [link](https://github.com/NVIDIA/spark-rapids/pull/9366/files#r1349396317)

Reply:
Okay lets keep it how it is because the conversion to/from pandas can be problematic.